### PR TITLE
[expo-cli] Add a build-status command

### PR DESCRIPTION
**Pull request by satya164**
# Why

We want to have a command to show the list of recent builds for turtle v2.

Example usage:

```js
expo build-status
```

### CODE REVIEW

File: packages/expo-cli/package.json [hunk 1]
```diff
@@ -64,7 +64,7 @@
   "dependencies": {
     "@expo/build-tools": "^0.1.10",
     "@expo/bunyan": "3.0.2",
-    "@expo/config": "3.2.15",
+    "@expo/config": "^3.2.15",
```
Comment (dsokal) [inline packages/expo-cli/package.json:67]:
```suggestion
    "@expo/config": "3.2.15",
```
Why this change? We're more or less trying to keep strict versions in package.json.

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 2]
```diff
@@ -70,12 +73,136 @@ export async function buildAction(projectDir: string, options: Options): Promise
   }
 }
 
+async function logBuildStatuses(
+  username: string,
+  buildStatus: {
+    jobs: Array<Record<string, any>>;
+    canPurchasePriorityBuilds: boolean;
+    numberOfRemainingPriorityBuilds: number;
+    hasUnlimitedPriorityBuilds?: boolean;
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:82]: For now, this only makes sense for turtle v1 builds.

Comment (dsokal) [reply packages/expo-cli/src/commands/build-native/index.ts:82]:
```suggestion
```

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 3]
```diff
@@ -70,12 +73,136 @@ export async function buildAction(projectDir: string, options: Options): Promise
   }
 }
 
+async function logBuildStatuses(
+  username: string,
+  buildStatus: {
+    jobs: Array<Record<string, any>>;
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:79]: 1. Rename `jobs` -> `builds`.
2. Add a type definition for builds - `Build` - and list all fields that we want to use in expo-cli.
3. Put the `Build` type definition in a separate file. I'm sure it'll be used for other code in `expo-cli` too.

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 4]
```diff
@@ -70,12 +73,136 @@ export async function buildAction(projectDir: string, options: Options): Promise
   }
 }
 
+async function logBuildStatuses(
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:76]: Please move the new functions to a separate file. We're trying to keep only the command definition in the command files.

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 5]
```diff
@@ -70,12 +73,136 @@ export async function buildAction(projectDir: string, options: Options): Promise
...
+    canPurchasePriorityBuilds: boolean;
+    numberOfRemainingPriorityBuilds: number;
+    hasUnlimitedPriorityBuilds?: boolean;
+  }
+) {
+  log('=================');
+  log(' Builds Statuses ');
+  log('=================\n');
+
+  buildStatus.jobs.forEach((job, i) => {
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:89]: We don't need to make the output of the new commands look exactly the same. IMO the output of the old command doesn't look great.
I think that we could use utils from `expo-cli/packages/expo-cli/src/commands/utils/cli-table.ts` to print a nice table with builds.
But feel free to propose sth.

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 6]
```diff
@@ -70,12 +73,136 @@ export async function buildAction(projectDir: string, options: Options): Promise
...
+      buildStatus.numberOfRemainingPriorityBuilds > 0 || buildStatus.hasUnlimitedPriorityBuilds;
+    const shouldShowUpgradeInfo =
+      !hasPriorityBuilds &&
+      i === 0 &&
+      job.priority === 'normal' &&
+      buildStatus.canPurchasePriorityBuilds;
+    let status;
+    switch (job.status) {
+      case 'pending':
+      case 'sent-to-queue':
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:116]: Builds v2 don't have the same build statuses as the old ones.
Take a look at this file - https://github.com/expo/universe/blob/master/server/www/src/constants.ts#L20

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 7]
```diff
@@ -70,12 +73,136 @@ export async function buildAction(projectDir: string, options: Options): Promise
...
+        status = `Build waiting in queue...\nQueue length: ${chalk.underline(
+          UrlUtils.constructTurtleStatusUrl()
+        )}`;
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:119]: Don't print the Turtle status url here. It's for Turtle v1.

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 8]
```diff
@@ -70,12 +73,136 @@ export async function buildAction(projectDir: string, options: Options): Promise
...
+        if (shouldShowUpgradeInfo) {
+          status += `\nWant to wait less? Get priority builds at ${chalk.underline(
+            'https://expo.io/settings/billing'
+          )}.`;
+        }
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:124]:
```suggestion
```

Comment (dsokal) [reply packages/expo-cli/src/commands/build-native/index.ts:124]: This is for Turtle v1 as well.

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 9]
```diff
@@ -88,7 +215,7 @@ export default function (program: Command) {
     .asyncActionProjectDir(buildAction, { checkConfig: true });
 
   program
-    .command('build:status')
+    .command('build-status [project-dir]')
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:218]: It could be worth adding the following options to filter the list:
- by platform - iOS or Android
- by build id - so we only get a build details output (rather than a list)

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 10]
```diff
@@ -70,12 +78,44 @@ export async function buildAction(projectDir: string, options: Options): Promise
...
+  });
+
+  const client = ApiV2.clientForUser(user);
+
+  const spinner = ora().start('Fetching build history...');
+
+  let buildStatus: { builds?: BuildInfo[] };
+
+  try {
+    buildStatus = await client.getAsync(`projects/${projectId}/builds/status`, {
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:100]:
```suggestion
    buildStatus = await client.getAsync(`projects/${projectId}/builds`, {
```

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 11]
```diff
@@ -70,12 +78,44 @@ export async function buildAction(projectDir: string, options: Options): Promise
...
+      platform: options.platform !== 'all' ? options.platform : undefined,
+      buildId: options.buildId,
+      status: options.status,
+    });
+  } catch (e) {
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:105]:
```suggestion
  } catch (e) {
    spinner.fail('... some reason ...');
```

File: packages/expo-cli/src/commands/build-native/index.ts (hunk 10 above)
Comment (dsokal) [reply packages/expo-cli/src/commands/build-native/index.ts:100]: Also, depending on what we decide to do with https://github.com/expo/universe/pull/5446#discussion_r459872658, we might need to detect if the user passed `--build-id` and handle this special case by making `GET /projects/PROJECT_ID/builds/BUILD_ID` request

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 12]
```diff
@@ -88,7 +128,10 @@ export default function (program: Command) {
     .asyncActionProjectDir(buildAction, { checkConfig: true });
 
   program
-    .command('build:status')
+    .command('build-status [project-dir]')
+    .option('-p --platform <platform>')
+    .option('-s --status <status>')
+    .option('-b --build-id <build-id>')
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:134]: Please add descriptions for these options (also, list possible values for enum options), sth like: https://github.com/expo/expo-cli/blob/master/packages/expo-cli/src/commands/upload.ts#L27

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 13]
```diff
@@ -70,12 +78,44 @@ export async function buildAction(projectDir: string, options: Options): Promise
...
+
+  const spinner = ora().start('Fetching build history...');
+
+  let buildStatus: { builds?: BuildInfo[] };
+
+  try {
+    buildStatus = await client.getAsync(`projects/${projectId}/builds/status`, {
+      platform: options.platform !== 'all' ? options.platform : undefined,
+      buildId: options.buildId,
+      status: options.status,
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/build-native/index.ts:103]:
```suggestion
      ...(['android', 'ios'].includes(platform) ? { platform } : {}),
      ...(buildId ? { buildId } : {}),
      ...(status ? { status } : {}),
```
otherwise you will endup with `?platform=&buildId=&status=`

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 14]
```diff
@@ -70,12 +78,57 @@ export async function buildAction(projectDir: string, options: Options): Promise
...
+  let builds: BuildInfo[] | undefined;
+
+  try {
+    const params = {
+      ...(['android', 'ios'].includes(platform) ? { platform } : null),
+      ...(status ? { status } : null),
+    };
+
+    if (buildId) {
+      const buildStatus = await client.getAsync(`projects/${projectId}/builds/${buildId}`, params);
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/build-native/index.ts:109]: this endpoint does not take those params
pair (status + platform) and build options should be mutually exclusive
verify that before sending request

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 15]
```diff
@@ -70,12 +85,77 @@ export async function buildAction(projectDir: string, options: Options): Promise
...
+  });
+
+  const client = ApiV2.clientForUser(user);
+
+  const spinner = ora().start('Fetching build history...');
+
+  let builds: BuildInfo[] | undefined;
+
+  try {
+    const params = {
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/build-native/index.ts:132]: if it's used only inside else, move this variable there or inline as function argument

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 16]
```diff
@@ -15,17 +17,30 @@ import AndroidBuilder from './AndroidBuilder';
 import iOSBuilder from './iOSBuilder';
 import { printBuildResults, printBuildTable, printLogsUrls } from './utils';
 
-interface Options {
-  platform: 'android' | 'ios' | 'all';
+const platforms = ['android', 'ios', 'all'] as const;
+const statuses = ['in-queue', 'in-progress', 'errored', 'finished'] as const;
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:21]: Hmm, maybe define that in the `build.ts` file and use `enum`:
```
enum BuildStatus {
  IN_QUEUE = 'in-queue',
  IN_PROGRESS = 'in-progress',
  ERRORED = 'errored',
  FINISHED = 'finished',
}
```

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 17]
```diff
@@ -15,17 +17,30 @@ import AndroidBuilder from './AndroidBuilder';
 import iOSBuilder from './iOSBuilder';
 import { printBuildResults, printBuildTable, printLogsUrls } from './utils';
 
-interface Options {
-  platform: 'android' | 'ios' | 'all';
+const platforms = ['android', 'ios', 'all'] as const;
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:20]: Maybe define an `enum` instead? 🤔 I haven't seen this syntax before.

File: packages/expo-cli/src/commands/build-native/index.ts [hunk 18]
```diff
@@ -15,17 +17,30 @@ import AndroidBuilder from './AndroidBuilder';
...
+const statuses = ['in-queue', 'in-progress', 'errored', 'finished'] as const;
+
+type BuildPlatform = typeof platforms[number];
+
+type BuildStatus = typeof statuses[number];
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/index.ts:25]: You won't need this after converting `platforms` and `statuses` to enums.

**Comment by dsokal** [comment]
Remember about updating the PR description :)

File: packages/expo-cli/src/commands/build-native/index.ts (hunk 17 above)
Comment (wkozyra95) [reply packages/expo-cli/src/commands/build-native/index.ts:20]: @dsokal I'm not sure if we want to use enums for those, especially platform type with all and without would be confusing

Comment (dsokal) [reply packages/expo-cli/src/commands/build-native/index.ts:20]: I'd still use an enum. We don't need to move it to any other file (like `build-native/build.ts`). In the context of this file, it'll be pretty straightforward what's the purpose of it (and why it has the `all` value included).

File: packages/expo-cli/src/commands/build-native/build.ts [hunk 19]
```diff
@@ -16,8 +16,15 @@ import log from '../../log';
 import { UploadType, uploadAsync } from '../../uploads';
 import { createProgressTracker } from '../utils/progress';
 
+export enum BuildStatus {
+  IN_QUEUE = 'in-queue',
+  IN_PROGRESS = 'in-progress',
+  ERRORED = 'errored',
```
Comment (dsokal) [inline packages/expo-cli/src/commands/build-native/build.ts:22]:
```suggestion
  ERRORED = 'errored',
  FINISHED = 'finished',
```

**Comment by wkozyra95** [comment]
@satya164 before merging make sure that all tests are passing
